lab/src/mapreduce/schedule.go

```diff
@@ -30,5 +33,42 @@ func schedule(jobName string, mapFiles []string, nReduce int, phase jobPhase, re
 	//
 	// Your code here (Part III, Part IV).
 	//
+
+	var workersPool sync.WaitGroup
+	var taskLock sync.Mutex
+
+	undoTask := ntasks
+	taskRemain := make(chan int, ntasks)
+	for taskNum := 0; taskNum < ntasks; taskNum++ {
+		taskRemain <- taskNum
+	}
+
+	for currentTaskNum := range taskRemain {
+		worker := <- registerChan
+		workersPool.Add(1)
+		go func(taskNum int, taskRemain chan int) {
+			ok := call(worker, "Worker.DoTask", DoTaskArgs{jobName, mapFiles[taskNum], phase, taskNum, n_other}, nil)
+			
+			defer workersPool.Done()
+			// register back to master
+			go func() {registerChan <- worker}()
```

Comment:
May I know why we want an additional thread to register the worker back?

Reply:
@Shauro If there is no consumer for this registerChan, the '<-' will be blocked which block the whole goroutine.

Here is the scenario:
When the last goroutine is launched and main thread is blocked in  line 46 waiting for new task or close() signal, no matter what happen in the last goroutine, call (close() first or register worker first(of course it will be blocked here)) there is no consumer for registerChan at this time, thus last goroutine will be blocked at 'registerChan <- worker' and main thread will be blocked at line 46 which is a dead lock there